test/Certificate.t.sol

```diff
@@ -109,9 +109,112 @@ contract Certificate_burn is UpgradeableCertificate {
       "Removal is not used for certificate"
     );
     assertEq(
-      _certificate.originalBalanceOf(0),
+      _certificate.purchaseAmount(0),
       1 ether,
       "Certificate original balance should be unchanged after burning"
     );
   }
 }
+
+contract Certificate_approve is UpgradeableCertificate {
+  function test() external {
+    vm.expectRevert(FunctionDisabled.selector);
+    _certificate.approve(_namedAccounts.buyer, 0);
+  }
+}
+
+contract Certificate_setApprovalForAll is UpgradeableCertificate {
+  function test() external {
+    vm.expectRevert(FunctionDisabled.selector);
+    _certificate.setApprovalForAll(_namedAccounts.buyer, true);
+  }
+}
+
+contract Certificate_safeTransferFrom is NonUpgradeableCertificate {
+  function setUp() external {
+    _mint(_namedAccounts.deployer, 1);
```

Comment:
it's a weird quirk of testing internal functions. AFAICT You can't actually `vm.prank` for internal calls to get it to work for the buyer since msgSender resolves to the deployer for the the reason I assume is explained by [this comment ](https://github.com/foundry-rs/foundry/issues/1454#issuecomment-1114011149)

Reply:
Not sure what you mean by this bit though `should this test be expected to succeed even if the certificate owner is not the account with the CERTIFICATE_OPERATOR_ROLE`

Reply:
```solidity

  function setUp() external {
    _mint(_namedAccounts.buyer, 1);
    _grantRole({
      role: CERTIFICATE_OPERATOR_ROLE,
      account: _namedAccounts.buyer
    });
  }

  function test() external {
    vm.prank(_namedAccounts.buyer);
    safeTransferFrom({
      from: _namedAccounts.buyer,
      to: _namedAccounts.buyer,
      tokenId: 0
    });
  }

```

**`[FAIL. Reason: TransferCallerNotOwnerNorApproved()]`**


```solidity
  function setUp() external {
    _mint(_namedAccounts.deployer, 1);
  }

  function test() external {
    safeTransferFrom({
      from: _namedAccounts.deployer,
      to: _namedAccounts.buyer,
      tokenId: 0
    });
  }
```
**[FAIL. Reason: ForbiddenTransferAfterMinting()]**

Reply:
in the case of pranking as the buyer, you end up with this in ERC721AUpgradeable's `transferFrom`
```
from--- 0xC6D4dE765734cb6B37FA07330bCFc241743eBbAf  // <-- buyer
address(uint160(prevOwnershipPacked)) != from)--- 0xC6D4dE765734cb6B37FA07330bCFc241743eBbAf // <-- buyer
_msgSenderERC721A()--- 0x465d5a3fFeA4CD109043499Fa576c3E16f918463 // <-- deployer :disappointed: 
```

Reply:
discussed offline in slack, takeaway being that the point of the `CERTIFICATE_OPERATOR_ROLE` is to mint our own certificates to our own admin account and then transfer them later.  should be the only role that can ever transfer their own tokens ever.